src/sync/channel.zig

```diff
@@ -1,170 +1,426 @@
+
+            fn create(allocator: Allocator) !*Buffer {
+                const new = try allocator.create(Buffer);
+                @memset(&new.slots, Slot.uninit);
+                new.next = Atomic(?*Buffer).init(null);
+                return new;
+            }
+
+            fn destroy(block: *Buffer, start: usize, allocator: Allocator) void {
+                for (start..BLOCK_CAP - 1) |i| {
+                    const slot = &block.slots[i];
+
+                    if (slot.state.load(.acquire) & READ == 0 and
+                        slot.state.fetchOr(DESTROY, .acq_rel) & READ == 0)
+                    {
+                        return;
+                    }
+                }
+
+                allocator.destroy(block);
+            }
+
+            fn deinit(block: *Buffer, allocator: Allocator) void {
+                if (block.next.load(.monotonic)) |n| {
+                    n.deinit(allocator);
+                    allocator.destroy(n);
+                }
+            }
+        };
+
+        const Slot = struct {
+            value: T,
+            state: Atomic(usize),
+
+            const uninit: Slot = .{
+                .value = undefined,
+                .state = Atomic(usize).init(0),
+            };
+        };
+
+        pub fn init(allocator: Allocator, initial_capacity: usize) Self {
+            _ = initial_capacity; // TODO: do something with this
+            return .{
+                .head = Position.init(),
+                .tail = Position.init(),
                 .allocator = allocator,
             };
-            return self;
         }
 
-        pub fn deinit(self: *Self) void {
-            var buff_lock = self.buffer.lock();
-            var buff: *std.ArrayList(T) = buff_lock.mut();
-            buff.deinit();
-            buff_lock.unlock();
+        pub fn create(allocator: Allocator, initial_capacity: usize) !*Self {
+            const channel = try allocator.create(Self);
+            channel.* = Self.init(allocator, initial_capacity);
+            return channel;
+        }
 
-            self.allocator.destroy(self);
+        pub fn send(channel: *Self, value: T) !void {
+            var backoff = Backoff.init();
+            var tail = channel.tail.index.load(.acquire);
+            var block = channel.tail.block.load(.acquire);
+            var next_block: ?*Buffer = null;
+
+            while (true) {
+                const offset = (tail >> SHIFT) % LAP;
+                if (offset == BLOCK_CAP) {
+                    backoff.snooze();
+                    tail = channel.tail.index.load(.acquire);
+                    block = channel.tail.block.load(.acquire);
+                    continue;
+                }
+
+                if (offset + 1 == BLOCK_CAP and next_block == null) {
+                    next_block = try Buffer.create(channel.allocator);
+                }
+
+                if (block == null) {
+                    const new = try Buffer.create(channel.allocator);
```

Comment:
Why not eagerly allocate the first block in Channel.init? Then this won't need to be nullable, and you can remove the null-handling logic from send and recv. It feels like a simplification that would help the readability/maintainability of the code.

Reply:
I added logic for creating the first block in the `init` function, good idea.